greenpithumb/dht11.py

```diff
@@ -0,0 +1,50 @@
+import datetime
+import threading
+
+from Adafruit_DHT import common
+
+_FRESHNESS_THRESHOLD = 2
+
+
+class CachingDHT11(object):
+    """Caching DHT11 sensor wrapper.
+
+    Reads and returns temperature and humidity levels while also caching these
+    values to ensure that the sensor is not polled at too high of a frequency.
+    """
+
+    def __init__(self, dht11_pin, clock):
+        """Creates a new CachingDHT11 object.
+
+        Args:
+            dht11_pin: An int representing the pin that the DHT11 sensor is
+                connected to.
+            clock: A clock interface
+        """
+        self._dht11_pin = dht11_pin
+        self._clock = clock
+        self._last_reading_time = datetime.min
+        self._last_reading = None
+        self._lock = threading.Lock()
+
+    def humidity(self):
+        """Returns a recent relative humidity reading."""
+        self._lock.acquire()
```

Comment:
Ooh, good thinking on making it thread safe!

Any time we can avoid explicitly calling `acquire` and `release` it's a win because it's much harder to make a mistake that leaves `acquire` and `release` mismatched, so suggest using

```python
with self._lock:
  ...
```

Also, I think the lock is slightly too broad. It doesn't hurt anything, but just so we're thinking deliberately about it, I'm trying to work through what the lock needs to protect. The `now = self._clock.now()` can go outside the lock because it's a local variable, so each thread has its own copy and there's no risk of overwriting values in the other thread's copies. The rest *does* need the lock because they deal with instances variables, which are shared across threads so other threads could clobber our values in the middle of execution if there's no lock.